### PR TITLE
fix: conditionally load MathJax

### PR DESCRIPTION
**Pull request by scissorsneedfoodtoo**
Removed MathJax CDN from header and set it up to download whenever a user goes to a Rosetta Code challenge

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #XXXXX

### CODE REVIEW

**Comment by scissorsneedfoodtoo** [comment]
Sorry, should have put this into draft mode. It's ready for review now.

**Comment by raisedadead** [comment]
> Sorry, should have put this into draft mode. It's ready for review now.

Resolve conflicts.

File: client/src/templates/Challenges/components/Side-Panel.js [hunk 1]
```diff
@@ -35,7 +33,8 @@ const propTypes = {
 };
 
 export class SidePanel extends Component {
-  componentDidMount() {
+  async componentDidMount() {
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/components/Side-Panel.js]:
```suggestion
  componentDidMount() {
```
I don't think this is necessary, see below.

File: client/src/templates/Challenges/components/Side-Panel.js [hunk 2]
```diff
@@ -35,7 +33,8 @@ const propTypes = {
...
+    const MathJax = await global.MathJax;
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/components/Side-Panel.js]:
```suggestion
    const MathJax = global.MathJax;
```
`MathJax` is just an ordinary object, no `await`ing required.

File: client/src/templates/Challenges/components/Side-Panel.js [hunk 3]
```diff
@@ -35,7 +33,8 @@ const propTypes = {
...
     if (MathJax) {
```
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/components/Side-Panel.js]: The case where MathJax isn't loaded needs to be handled, too.  This https://github.com/freeCodeCamp/freeCodeCamp/blob/master/client/src/components/Donation/components/DonateModal.js#L55-L61 is how it was handled for Stripe.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Hey @scissorsneedfoodtoo once you've visited a challenge everything looks good - it's just that typically MathJax isn't going to be ready when the component mounts.

**Comment by scissorsneedfoodtoo** [comment]
Will wait for https://github.com/freeCodeCamp/freeCodeCamp/pull/37388 to land before updating further.

**Comment by raisedadead** [comment]
@scissorsneedfoodtoo bump for further work.

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams, thank you for your earlier review. I went through and made the changes you suggested.

I updated the MathJax fallback script loader to work like the one in `donate.js`, but it always seems to run for some reason. Took a look at `gatsby-ssr.js` and can't figure out while the Stripe script loads when on a challenge or `/donate` but not on any of the Rosetta Code problems, leading to the fallback running when users first visit a Rosetta Code challenge.

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo I think the issue is that https://github.com/freeCodeCamp/freeCodeCamp/blob/4ff2870ced8b67ad98a81efdf6a5fa5d9c88171d/client/gatsby-ssr.js#L66
should not have a trailing slash, since the `pathname`s Gatsby returns don't have them.

**Comment by ojeytonwilliams** [comment]
So, yeah, sorry about that!  I think that was my mistake.

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams, no need to apologize! I should have thought of trying that. Thank you for pointing that out -- learned something new about Gatsby.

I'll go in and remove that slash.

**Comment by ojeytonwilliams** [comment]
Thanks for sorting that out!

So, I've been having a look at it in development mode and I noticed something odd.  This is what you get when you reload the first challenge (100 doors)

image
while on .dev it looks like this
image

But, if you navigate to 100 doors from another challenge, it's fine.

I think the configuration is getting messed up somehow (it's treating `(` as delimiters, but only some of the time).  I'll see if I can figure out why after I grab some lunch.

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams, great catch. I've been looking at the very math heavy pages and didn't notice that at all.

I believe you're right, MathJax is treating `(` as delimiters. Probably has to do with the way I'm trying to set the configuration in the fallback script in `scriptLoaders.js`. The parentheses probably need to be escaped there, but we may be able to take that extra configuration out entirely.

I'll do some quick testing.

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo I'm glad you realised they needed escaping twice, I had not noticed there were some challenges that used them. I was going to suggest we cut them out.  Oops.

Anyway, I've added some changes to simplify things a bit.  I realised that using `state` as an intermediary wasn't necessary and we could just see if `global.MathJax` was defined or not.  I'm going to double check this doesn't break anything, but I think it's okay.

**Comment by scissorsneedfoodtoo** [comment]
@ojeytonwilliams, awesome! Thank you for simplifying this much further. It's much easier to reason about now.

I also thought it would be better to remove the backticks until I reviewed the MathJax documentation again. Doesn't look great now, but I'm glad it's working as intended.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
@scissorsneedfoodtoo if you're happy I think we're basically there. However, since I've been tweaking the code @ahmadabdolsaheb could give this a quick review?

**Comment by ahmadabdolsaheb** [comment]
tested locally, LGTM

**Pull request merged**